### ISSUE TITLE
Colyseus 0.14 ref id not found

### ISSUE THREAD

**Issue by doorbash** [colyseus/colyseus#369]
Hello

# 1

Sending a `Schema` object to client which is part of the room state causes this error
but it's not happening when sending a new object:

## Example:
This causes error:

```typescript
onCreate (options: any) {
  var state = new MyRoomState()
  state.player = new Player()
  this.setState(state);

  this.clock.setTimeout(() => {
    this.broadcast(state.player)
  }, 3000)
}
```

This doesn't:
```typescript
onCreate (options: any) {
  var state = new MyRoomState()
  state.player = new Player()
  this.setState(state);

  this.clock.setTimeout(() => {
    this.broadcast(new Player())
  }, 3000)
}
```

## Steps to reproduce:
1. Run this: https://github.com/doorbash/colyseus-ref-id-not-found
2. Connect with any client and wait 3s

# 2

This one is reported to me by @vinsmain via email and I ran it with `js` client and fortunately it happened there too!

## Steps to reproduce:
1. Run this: https://github.com/doorbash/colyseus-ref-id-not-found2
2. Connect to it like this:
```typescript
import { Client } from ".";

const client = new Client("ws://localhost:2567");

(async () => {
    try {
        const room = await client.joinOrCreate<any>("game");

        room.state.selectedInvestigators.onAdd = ((value, key) => {
            console.log("ADD INV " + key + " " + value.influence);

            if (key == 0) room.send("inv", "null");
        });

        room.send("addCard")

    } catch (e) {
        console.error("couldn't join room:", e);
    }
})()
```

**Comment by endel** [comment]
Thanks a lot for reporting @doorbash ❤️ I'm in the process of releasing Colyseus 0.14 fixing this right now! 🎉

**Comment by doorbash** [comment]
That's great news @endel Thank you!

**Comment by vinsmain** [comment]
Hello @endel

Server version 0.14.1

The problem with decreasing the length of the array is solved, but if
you increase the length (push) and shuffle - the same error.

> Thanks a lot for reporting @doorbash ❤️ I'm in the process of releasing Colyseus 0.14 fixing this right now! 🎉

**Comment by endel** [comment]
Hi @vinsmain, could you provide a minimal way to reproduce the error? How does your shuffle method look like?

**Comment by vinsmain** [comment]
> Hi @vinsmain, could you provide a minimal way to reproduce the error? How does your shuffle method look like?

The test code is the same, only the server is updated to 0.14.1 and schema to 1.0.0
https://github.com/doorbash/colyseus-ref-id-not-found2/blob/master/model/GameServer.ts
Function shuffle on line 162.

**Comment by endel** [comment]
Hi @vinsmain, I've created a PR for you and @doorbash here: https://github.com/doorbash/colyseus-ref-id-not-found2/pull/1

Could you modify the `colyseus-ref-id-not-found2.ts` file in a way that I can see the error happening? You can run `npm test` to execute that file.

Basically, after you perform some mutations in the `state`, you should call `decodedState.decode(state.encode())` to simulate those changes being transferred to the client.

Let me know if you manage to reproduce it there!

**Comment by vinsmain** [comment]
Hi @doorbash, as able to reproduce as follows

https://gist.github.com/vinsmain/a44b34ed2a028fcfccafeeeef98298e8

![image](https://user-images.githubusercontent.com/20394400/96910947-da08c080-14a8-11eb-9028-e55be04b0870.png)

**Comment by doorbash** [comment]
@vinsmain I've just added your gist to the PR

**Comment by endel** [comment]
Thanks @vinsmain @doorbash, I'm inspecting this right now, looks like something wrong during encoding

**Comment by endel** [comment]
Hey @vinsmain, just published `@colyseus/schema@1.0.1` fixing this! Let me know if you face any other issues!
